core: RPMB (Replay Protected Memory Block) filesystem support

jforissier opened the pull request:
This is an update to the RPMB filesystem implementation so that the
persistent object API may use RPMB rather than the REE filesystem.
This feature is enabled with CFG_RPMB_FS=y.
Note that this implementation requires support from the non-secure side
to actually access the RPMB partition, as there is no eMMC driver here.
Also, the code is currently not compatible with CFG_ENC_FS (file
encryption), which must be set to 'n'. Encryption will be added later.

** Note: Initial patches from:
   https://github.com/Microsoft/optee_os/tree/rpmb2
Youssef/Paul, please let me know if you're OK with having your
Signed-off-by: below. Also it would be good if you could provide a
Tested-by: for your platform, or discuss any issue you might have with
the changes I have introduced. I will delete this notice before the
patch is merged. **

Signed-off-by: Youssef Esmat <youssef.esmat@microsoft.com>
Signed-off-by: Paul Swan <Paul.Swan@microsoft.com>
[Rebased onto master, Linux driver/tee-supplicant support]
Signed-off-by: Jerome Forissier <jerome.forissier@linaro.org>

jenswi-linaro commented on core/arch/arm/tee/tee_rpmb.c:
```diff
@@ -162,30 +163,35 @@ struct tee_rpmb_ctx {
-{
-	return crypto_ops.hash.init(ctx, TEE_ALG_HMAC_SHA256);
-}
+static uint8_t rpmb_test_key[RPMB_KEY_MAC_SIZE] = {
```
please make it `const` also.

jforissier replied:
OK

jenswi-linaro commented on core/arch/arm/tee/tee_rpmb.c:
```diff
@@ -305,20 +333,47 @@ static TEE_Result tee_rpmb_mac_calc(uint8_t *mac, uint32_t macsize,
 	size_t resp_size;
 };
 
+static void tee_rpmb_clear_session(struct tee_ta_session **session)
```
These aren't needed any longer for RPC now that the user TA translation tables are assigned per thread instead of per CPU as it was in the past.

jforissier replied:
Indeed, that is neat. Will fix.

jenswi-linaro commented on core/tee/tee_rpmb_fs.c:
```diff
@@ -90,18 +97,109 @@ struct rpmb_fs_partition {
+ * pointer to name here.
+ */
+struct tee_rpmb_fs_dirent {
+	struct tee_rpmb_fs_dirent *next;
```
I would prefer if we where using the queue.h macros for linked lists.

jforissier replied:
OK, I have replaced this with a `SIMPLEQ_HEAD`.

jenswi-linaro commented on core/arch/arm/tee/tee_rpmb.c:
```diff
@@ -804,18 +863,36 @@ static TEE_Result tee_rpmb_init_read_wr_cnt(uint16_t dev_id,
-	uint16_t op_result;
 
-	if (rpmb_ctx == NULL) {
+	(void)writekey;
```
Please use the `__unused` tag in the function declaration instead.

jforissier replied:
In fact I can probably just drop `writekey` and `commercial`, they're not used and not part of an external API either.

jenswi-linaro commented on core/arch/arm/tee/tee_rpmb.c:
```diff
@@ -804,18 +863,36 @@ static TEE_Result tee_rpmb_init_read_wr_cnt(uint16_t dev_id,
 static TEE_Result tee_rpmb_init(uint16_t dev_id, bool writekey, bool commercial)
 {
-	TEE_Result res = TEE_ERROR_GENERIC;
+	TEE_Result res = TEE_SUCCESS;
```
We usually check the presence of relevant crypto functions before using them, somewhere in this function would be convenient to do those checks.

jforissier replied:
Yes.

jenswi-linaro commented on core/arch/arm/tee/tee_rpmb.c:
```diff
@@ -215,30 +221,52 @@ static TEE_Result tee_get_hw_unique_key(struct tee_hw_unique_key *hwkey)
+		goto out;
+
+	res = crypto_ops.mac.final(ctx, TEE_ALG_HMAC_SHA256, key, len);
+	if (res != TEE_SUCCESS)
```
This check doesn't do anything.

jforissier replied:
I'll remove it.

jenswi-linaro commented on core/arch/arm/tee/tee_rpmb.c:
```diff
@@ -305,9 +333,12 @@ static TEE_Result tee_rpmb_mac_calc(uint8_t *mac, uint32_t macsize,
 
 struct tee_rpmb_mem {
-	paddr_t phpayload;
+	struct teesmc32_arg *arg;
```
This isn't used in the RPC communication.

jforissier replied:
OK, I suppose I just need to allocate a `struct teesmc32_param params[2]` on the stack in `tee_rpmb_invoke()`, similar to what is done by `wq_rpc()` in `wait_queue.c` for instance?

jforissier commented:
Updated.

jenswi-linaro commented on core/tee/tee_rpmb_fs.c:
```diff
@@ -26,21 +26,26 @@
 #define RPMB_FS_MAGIC                   0x52504D42
-#define FS_VERSION                      1
-#define N_ENTRIES                       16
+#define FS_VERSION                      2 /* 1 */
```
That comment isn't very useful, same below.

jforissier replied:
Will fix.

pascal-brand-st-dev commented on core/tee/tee_rpmb_fs.c:
```diff
@@ -90,18 +98,109 @@ struct rpmb_fs_partition {
-		fh->filename = filename;
+	if (filename) {
+		len = strlen(filename);
+		memcpy(fh->filename, filename, len);
```
strncpy instead

jforissier replied:
`strlcpy()` you mean? I will use that.

pascal-brand-st-dev commented on core/tee/tee_rpmb_fs.c:
```diff
@@ -149,10 +254,22 @@ static TEE_Result rpmb_fs_setup(void)
+		goto out;
+	}
+
+	DMSG("");
```
There are many traces. Can some be cleaned?

jforissier replied:
Yes, I agree that many of them are not very useful. I'll remove them.

pascal-brand-st-dev commented on core/tee/tee_rpmb_fs_common.c:
```diff
@@ -0,0 +1,318 @@
+{
+	int res = -1;
+
+	if (fdp == NULL)
```
`if (!fdp)`

jforissier replied:
OK

pascal-brand-st-dev commented on core/arch/arm/tee/tee_rpmb.c:
```diff
@@ -213,32 +218,48 @@ static TEE_Result tee_get_hw_unique_key(struct tee_hw_unique_key *hwkey)
+				   uint8_t *key, uint32_t len)
 {
-	TEE_Result res = TEE_ERROR_GENERIC;
+	TEE_Result res = TEE_SUCCESS;
```
No need to initialize

jforissier replied:
OK

pascal-brand-st-dev commented on core/arch/arm/tee/tee_rpmb.c:
```diff
@@ -804,18 +841,42 @@ static TEE_Result tee_rpmb_init_read_wr_cnt(uint16_t dev_id,
+	if (!have_crypto_ops())
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	if (!rpmb_ctx) {
```
`rpmb_ctx` is global. It should be protected by a mutex in case of concurrent TA. It could be done later on in another PR.

jforissier replied:
Yes, I can add the mutex to this patch, I think it's better.

pascal-brand-st-dev commented:
We're also missing a `.md` file. It can be done in another PR.

jenswi-linaro commented on core/tee/tee_rpmb_fs.c:
```diff
@@ -228,16 +353,12 @@ static TEE_Result get_fat_start_address(uint32_t *addr)
 {
 	TEE_Result res = TEE_ERROR_GENERIC;
```
This function could be simplified by not using this variable.

jforissier replied:
Agreed

jenswi-linaro commented on core/tee/tee_rpmb_fs_common.c:
```diff
@@ -0,0 +1,318 @@
+	}
+
+	fd = res;
+	fdp = (struct tee_fs_fd *)malloc(sizeof(struct tee_fs_fd));
```
Cast is not needed.

jforissier replied:
OK

jenswi-linaro commented on core/tee/tee_rpmb_fs_common.c:
```diff
@@ -0,0 +1,318 @@
+	fdp->nw_fd = fd;
+	fdp->flags = flags;
+	fdp->is_new_file = is_new_file;
+	fdp->filename = malloc(len);
```
we have `strdup()` for this.

jforissier replied:
Yes

jenswi-linaro commented on core/tee/tee_rpmb_fs_common.c:
```diff
@@ -0,0 +1,318 @@
+}
+
+
+static TEE_Result _to_errno(int rc)
```
The leading underscore shouldn't be needed.

jforissier replied:
Right, will fix

jforissier commented:
Updated.

jenswi-linaro commented:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jforissier commented:
I think I have addressed all the comments, @pascal-brand-st-dev are you OK with this PR?
Regarding documentation: I will make another patch on top of this and https://github.com/OP-TEE/optee_os/pull/626 so that I can document everything including encryption.

pascal-brand-st-dev commented:
Reviewed-by: Pascal Brand <pascal.brand@linaro.org>

jforissier commented:
Rebased and tags applied, thanks for reviewing this big PR!

The Travis build appears to be broken due to an unrelated commit, which should be fixed by https://github.com/OP-TEE/build/pull/42, and also because `checkpatch.pl` is rejecting the use of `SIMPLEQ_HEAD` -- which is perfectly fine I think.
